### DISCUSSION TITLE
Support indirect dependency (lockfile) updates for vulnerability alerts

### DISCUSSION THREAD

**Discussion by jalaziz** [renovatebot/renovate#28464] [Suggest an Idea]
### Tell us more.

I know this is not currently supported, but it would be great if Renovate could update indirect dependencies in lock files when a vulnerability alert is triggered against them.

While this can be indirectly handled with `lockFileMaintenance`, it's a pretty heavy hammer when a targeted update would be better for understanding why the update is important.

In our case, the issue is with the `mio` dependency that is an indirect dependency for `tokio`. [This `tokio` issue](https://github.com/tokio-rs/tokio/issues/6369#issuecomment-1978292853) helps explain why this would be useful.

**Comment by Sebazzz** [comment]
That would require some complex logic, because sometimes it doesn't help or isn't possible to do a deep update and you need to pin the transitive dependency explicitly in the `package.json` using `resolutions`.

**Reply by jalaziz** [reply]
For Rust/Cargo, you can simply call `cargo update <dep>` and it will update the `Cargo.lock` based on the constraints of the workspace.

I'd imagine it's nearly the same as `update-lockfile` but targeting indirect dependencies. I realize I'm trivializing the problem a bit, but seems like (at least in Rust's case), `cargo` can handle most of the heavy lifting.

It certainly won't be possible to always solve the vulnerability alert based on the dependency versions in `Cargo.toml`, but it seems like a decent number of pieces are in place between the `update-lockfile` strategy and lockfile maintenance.

The part I'm not so sure about is whether Renovate has any general concept of extracting and updating indirect dependencies.

I doubt this is something that can be done universally, but there's already precedent for some features being supported by a subset of manager.

**Reply by joevin-slq-docto** [reply]
Dependabot does it without any issue on multiple packages managers, at least with Bundler and NPM.
<details><summary>Example</summary>
<p>
From a Dependabot Alert, Dependabot can create a pull request that will modify the Gemfile.lock to only bump the transitive dependency :
<img width="766" height="197" alt="image" src="https://github.com/user-attachments/assets/96e7158e-3de3-494b-9b0c-471c74bc5427" />

As explained jalaziz, `lockFileMaintenance` is not a solution as it is too unreliable to bump all transitive in one pull request.
</p>
</details>
